src/main/java/org/jenkinsci/plugins/pipeline/milestone/MilestoneStepExecution.java

```diff
@@ -0,0 +1,432 @@
+import java.util.logging.Logger;
+
+import javax.annotation.CheckForNull;
+
+import org.jenkinsci.plugins.workflow.actions.LabelAction;
+import org.jenkinsci.plugins.workflow.actions.ThreadNameAction;
+import org.jenkinsci.plugins.workflow.flow.FlowExecutionOwner;
+import org.jenkinsci.plugins.workflow.graph.BlockEndNode;
+import org.jenkinsci.plugins.workflow.graph.BlockStartNode;
+import org.jenkinsci.plugins.workflow.graph.FlowGraphWalker;
+import org.jenkinsci.plugins.workflow.graph.FlowNode;
+import org.jenkinsci.plugins.workflow.graph.FlowStartNode;
+import org.jenkinsci.plugins.workflow.steps.AbstractSynchronousStepExecution;
+import org.jenkinsci.plugins.workflow.steps.FlowInterruptedException;
+import org.jenkinsci.plugins.workflow.steps.StepContext;
+import org.jenkinsci.plugins.workflow.steps.StepContextParameter;
+
+import com.google.inject.Inject;
+
+import hudson.AbortException;
+import hudson.Extension;
+import hudson.XmlFile;
+import hudson.model.Executor;
+import hudson.model.InvisibleAction;
+import hudson.model.Job;
+import hudson.model.Result;
+import hudson.model.Run;
+import hudson.model.TaskListener;
+import hudson.model.listeners.RunListener;
+import jenkins.model.CauseOfInterruption;
+import jenkins.model.Jenkins;
+
+public class MilestoneStepExecution extends AbstractSynchronousStepExecution<Void> {
+
+    private static final Logger LOGGER = Logger.getLogger(MilestoneStepExecution.class.getName());
+
+    @Inject(optional=true) private transient MilestoneStep step;
+    @StepContextParameter private transient Run<?,?> run;
+    @StepContextParameter private transient FlowNode node;
+    @StepContextParameter private transient TaskListener listener;
+
+    @Override
+    public Void run() throws Exception {
+        if (step.getLabel() != null) {
+            node.addAction(new LabelAction(step.getLabel()));
+        }
+        int ordinal = processOrdinal();
+        tryToPass(run, getContext(), ordinal);
+        return null;
+    }
+
+    /**
+     * Gets the next ordinal and throw {@link AbortException} the milestone lives inside a parallel step branch.
+     */
+    private synchronized int processOrdinal() throws AbortException {
+        FlowGraphWalker walker = new FlowGraphWalker();
+        walker.addHead(node);
+        Integer previousOrdinal = null;
+        int parallelDetectionEnabled = 0;
+        for (FlowNode n : walker) {
+
+            if (parallelDetectionEnabled <= 0 && n.getAction(ThreadNameAction.class) != null) {
+                listener.getLogger().println("Milestone step found inside parallel, it's not possible to grant ordering in this case.");
+                throw new AbortException("Using a milestone step inside parallel is not allowed");
+            }
+
+            if (n instanceof BlockEndNode) {
+                parallelDetectionEnabled++;
+            } else if (n instanceof BlockStartNode && !(n instanceof FlowStartNode)) {
+                parallelDetectionEnabled--;
```

Comment:
`FlowNodeSerialWalker` would not work here I think because of:

```groovy
node {
  milestone()
}
node {
  milestone()
}
```

From the second `milestone`, we need to find the first, which is not “up and back”.